### ISSUE TITLE
Better performance with large datasets

### ISSUE THREAD

**Issue by bbyars** [bbyars/mountebank#279]
See https://groups.google.com/forum/#!topic/mountebank-discuss/RQf2bhwCHLY

Large data sets (e.g. 1000+ request/response pairs) can slow mountebank response time, presumably due to the memory footprint and serial execution of predicate tests. Examine ways of improving performance, possibly including:

* Storing state on disk rather than in memory for large datasets
* Parallel predicate execution

**Comment by bbyars** [comment]
Thinking through solutions, I believe this is closely related to #382 and #167 and, indirectly, #177.

Initial thoughts:
My initial reaction was to reach for a database, but I don't think that adds anything. The query capability is unique to mountebank (especially with things like jsonpath predicates). If anyone has a counterargument, please let me know. I'm more than willing to add infrastructure, but if it doesn't add any value, simply using the filesystem makes installation much simpler.

Keeping everything in memory is clearly a scaling problem. The solution likely involves a robust file layout, which adds persistence (#167) and eases test data management (#177). 

Still evolving solution ideas over a bottle of gin, but I believe a filesystem layout that acts like a search tree is the right solution. Initial ideas

````
imposters.json
3000.json  (header information)
8080.json
3000/   (the imposter port)
    requests/  (when --mock is used)
        0.json  (first request)
        1.json
    stubs/
        0-99.json  (load 100 stubs at a time)
        100-199.json
        responses/
            0[0].json  (first response of first stub)
            0[1].json   (second response of first stub)
            1[0].json   (first response of second stub)
8080/
````

Saving requests can no longer be a memory leak, just write it to `numberOfRequests`.json (and mountebank already maintains state of `numberOfRequests`). I'd still only do it with the --mock flag because you can maximize performance without it. 

Checking a request would load the first 100 stubs, and if not stub's predicates match the incoming request, load the second 100, and so on. I believe serial execution of predicate matching is still right, but open to suggestions.

For clean backwards compatibility, a parallel set of .ejs files can be created that would load with --configfile.

**Comment by bbyars** [comment]
After thinking on it some more, The `responses` filenames should not indicate order, as that would incur too much overhead during recording to reshuffle filenames, especially for large datasets. Instead, use perhaps a sequence identifier for the filename and another file that maps the order to the filename. Something like

````
stubs/
    responseIndex.json
    responses/
        0.json
        1.json
````

**Comment by colinschoen** [comment]
The hierarchical file structure to serialize stubs you've described makes sense from a persistence and data management perspective, but I'm not sure memory is the bottleneck. From my understanding, the search tree with the serialized stubs design will still be the same size as the status quo. 

Let me give a bit more background on our use case and how we ran into the scaling problem:

We are using Mountebank to facilitate [VCR testing](https://www.youtube.com/watch?v=wJPfsyRRmyw&index=26&list=PLSIUOFhnxEiCWGsN9t5A-XOhRbmz54IS1) as a network level proxy to record requests to service dependencies in "record mode" and replay them back in what we are calling "replay mode". We have an internal playground/sandbox management tool similar to Docker Compose. We have built functionality directly in this tool to automatically generate a Mountebank configuration from dependency information (imposters for each service whose interactions are being recorded) to be used in "record mode" . We will run the test suite and the playground tool will communicate with mountebank, fetch the stubs and serialize the recorded stubs to disk for use in replay mode. We separate these serialized stubs by service and test (we wrote a pytest plugin to monkey patch python HTTP libraries to inject a header containing the test for which a request has provenance to. This header is propagated by requests made by the system under test and other services throughout the request life.). 

We were running into performance issues as we had 500+ plus stubs per imposter in replay mode that would have to be searched through even though there were at most 20-30 stubs for a specific service during one test. Since we are generating predicates that match on a specific header containing the test_id we were able to reduce the search space by essentially using mountebank as an intelligent load balancer for itself. 

The mountebank configuration automatically generated in replay mode by the sandbox tool was modified to now create a high level imposter for each service being faked. This high level imposter contains stubs which simply match on the one header that contains the test_id and then serves a proxy response to a specific imposter for that service & test_id. We saw performance improvements upwards of 80%.

This is a unique use case, but I do think being able to index on a specific predicate would have been useful here and in other cases too. 

Tangent over... :) 

I can do some testing to see if memory is in fact the bottleneck, but I think the problem may just persist unless the search tree is shrunken or Mountebank uses a webserver with multiple workers or [cluster](https://nodejs.org/api/cluster.html#cluster_cluster) etc
